### ISSUE TITLE
Invalid CS failure for `PSR12.Functions.NullableTypeDeclaration.UnexpectedCharactersFound`

### ISSUE THREAD

**Issue by ScopeyNZ** [squizlabs/PHP_CodeSniffer#2552]
### Affected version

3.x-dev (dev-master)

### Description

With a valid piece of code I'm managing to get a `PSR12.Functions.NullableTypeDeclaration.UnexpectedCharactersFound` error, on a line that's not even a function declaration.

PHP File:

```php 
<?php
namespace {
    class Test
    {
        public function test()
        {
            $test = Something::one(static::CONSTANT) ?: '';
        }
    }
}
```

PHPCS config:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="test">
    <description>Test standards</description>

    <!-- base rules are PSR-12 -->
    <rule ref="PSR12" >
    </rule>
</ruleset>
```

**Expected result**: Nothing (valid) 🙂 

**Actual result**:

```
gmarriott@SSguyMBP:[~/tmp/phpcs-sandbox]: phpcs test.php -s

FILE: /Users/gmarriott/tmp/phpcs-sandbox/test.php
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 7 | ERROR | There must not be a space between the question mark and the type in nullable type declarations
   |       | (PSR12.Functions.NullableTypeDeclaration.UnexpectedCharactersFound)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Time: 53ms; Memory: 6MB
```

### Notes:

It's related to the constant call. Switching to `self::CONSTANT` actually fixes the issue - despite the fact the code probably won't work then as the constant isn't defined. It doesn't matter if the constant is defined or not, it will fail in either case (with `static`).

**Comment by gsherwood** [comment]
Issue is with the tokenizer, which thinks the `?` is a `T_NULLABLE`. Needs to be fixed there.

**Comment by jrfnl** [comment]
@gsherwood I've got a fix ready for this.

**Comment by gsherwood** [comment]
@jrfnl Thanks for fixing this. I couldn't replicate on the current stable version so I haven't added a changelog entry for it. If this is a mistake, please let me know and I'll adjust accordingly.

**Comment by jrfnl** [comment]
> I couldn't replicate on the current stable version

Just checked, but without this fix the issue was still reproducible.

Just check out the commit before the merge, cherrypick the unit test file change and run PHPCS with `-vv` to see that the `?` on line 65 would be incorrectly changed to `T_NULLABLE`. With the fix in, it no longer is.

**Comment by gsherwood** [comment]
> Just check out the commit before the merge, cherrypick the unit test file change and run PHPCS with `-vv` to see that the `?` on line 65 would be incorrectly changed to `T_NULLABLE`. With the fix in, it no longer is.

Sorry, I meant I couldn't replicate on 3.4.2. The `?` tokenizes as `T_INLINE_THEN` for me on 3.4.2, but it was easily reproducible on master before your fix.

**Comment by jrfnl** [comment]
Ah, I see what you mean. Looks like the issue was introduced by 100a81ca6236dfca27f5d18628ce607dd673c0cc in that case.

**Comment by gsherwood** [comment]
> Looks like the issue was introduced by [100a81c](https://github.com/squizlabs/PHP_CodeSniffer/commit/100a81ca6236dfca27f5d18628ce607dd673c0cc) in that case.

That would be the one!

**Comment by tehbeard** [comment]
I have ran into this bug while switching to PSR12 (running 3.5.0).
It seems to be related to the use of `new static(...)`

Console output:
```
PS G:\phpcs-bug> phpcs --version
PHP_CodeSniffer version 3.5.0 (stable) by Squiz (http://www.squiz.net)

PS G:\phpcs-bug> phpcs -s --standard=psr12 --report=code --report-width=120 .\Attachment.php -v
Registering sniffs in the PSR12 standard... DONE (59 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory G:\phpcs-bug
Processing Attachment.php [PHP => 282 tokens in 42 lines]... DONE in 10ms (2 errors, 0 warnings)

FILE: G:\phpcs-bug\Attachment.php
------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
------------------------------------------------------------------------------------------------------------------------
LINE 34: ERROR [x] There must not be a space between the question mark and the type in nullable type declarations
                   (PSR12.Functions.NullableTypeDeclaration.WhitespaceFound)
------------------------------------------------------------------------------------------------------------------------
   32:          $finfo = new \finfo(FILEINFO_MIME_TYPE);
   33:          return new static(
>> 34:              (is_null($name) ? basename($file) : $name),
   35:              is_null($type) ? $finfo->file($file) : $type,
   36:              null,
------------------------------------------------------------------------------------------------------------------------
LINE 35: ERROR [ ] There must not be a space between the question mark and the type in nullable type declarations
                   (PSR12.Functions.NullableTypeDeclaration.UnexpectedCharactersFound)
------------------------------------------------------------------------------------------------------------------------
   33:          return new static(
   34:              (is_null($name) ? basename($file) : $name),
>> 35:              is_null($type) ? $finfo->file($file) : $type,
   36:              null,
   37:              $file
------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------------------------------------------------
Time: 278ms; Memory: 4MB
```

PHP file:
```php
<?php

namespace rebar\mail\letter;

class Attachment
{
    private $name;
    private $type;
    private $contents;
    private $file;

    /**
     * Attachment constructor.
     *
     * @param $name
     * @param $type
     * @param $contents
     * @param $file
     */
    private function __construct($name, $type, $contents, $file)
    {
        $this->name = $name;
        $this->type = $type;
        $this->contents = $contents;
        $this->file = $file;
    }

    public static function fromFile($file, $name = null, $type = null)
    {
        $test = Something::one(static::CONSTANT) ?: ''; //Added to test if regression of previous test case

        $finfo = new \finfo(FILEINFO_MIME_TYPE);
        return new static(
            (is_null($name) ? basename($file) : $name),
            is_null($type) ? $finfo->file($file) : $type,
            null,
            $file
        );
    }

    public static function fromFileUsingSelf($file, $name = null, $type = null)
    {
        //Using self instead of static
        $test = Something::one(static::CONSTANT) ?: ''; //Added to test if regression of previous test case

        $finfo = new \finfo(FILEINFO_MIME_TYPE);
        return new self(
            (is_null($name) ? basename($file) : $name),
            is_null($type) ? $finfo->file($file) : $type,
            null,
            $file
        );
    }

    // rest of methods removed for test
}

```
